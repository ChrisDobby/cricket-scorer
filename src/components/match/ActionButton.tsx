import * as React from 'react';

const buttonStyle: React.CSSProperties = {
    minWidth: '36px',
    margin: '2px',
};

export interface ActionButtonProps {
    caption: string;
    noBall: boolean;
    action: () => void;
}

export const ActionButton = ({ caption, noBall, action }: ActionButtonProps) => (
    <button
        className={`btn ${noBall ? 'btn-danger' : 'btn-success'}`}
        style={buttonStyle}
        onClick={action}
    >{caption}
    </button>
);
