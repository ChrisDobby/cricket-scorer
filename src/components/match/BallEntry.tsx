import * as React from 'react';
import * as globalStyles from '../styles';
import { Innings, Batter, Bowler, oversDescription, BattingInnings } from '../../domain';
import { EntryPanel, BallFunctions } from './EntryPanel';

const battingScoreStyle: React.CSSProperties = {
    display: 'inline',
    marginRight: '10px',
};

export interface BallEntryProps {
    innings: Innings;
    batter: Batter;
    bowler: Bowler;
    ballFunctions: BallFunctions;
}

export const BallEntry = ({ innings, batter, bowler, ballFunctions }: BallEntryProps) => (
    <div style={globalStyles.sectionContainer}>
        <div className="row" style={globalStyles.singleHeadingRow}>
            <h4>{`${bowler.name} to ${batter.name}`}</h4>
        </div>
        <div className="row">
            <div className="col-12 col-lg-6">
                <div className="row">
                    <div className="col-6">
                        <h5>{innings.battingTeam.name}</h5>
                    </div>
                    <div className="col-6">
                        <h5>{`${innings.score}-${innings.wickets}`}</h5>
                    </div>
                </div>
                <div className="row">
                    <div className="col-6">
                        <h5>Overs</h5>
                    </div>
                    <div className="col-6">
                        <h5>{oversDescription(innings.completedOvers, [])}</h5>
                    </div>
                </div>
                {innings.batting.batters.filter(batter => batter.innings && !batter.innings.wicket)
                    .map((batter, idx) => ({
                        name: batter.name,
                        innings: batter.innings as BattingInnings,
                        index: idx,
                    }))
                    .map(currentBatter => (
                        <div key={currentBatter.index} className="row">
                            <div className="col-6">
                                <h5>{currentBatter.name}</h5>
                            </div>
                            <div className="col-6">
                                <span>
                                    <h5 style={battingScoreStyle}>{`${currentBatter.innings.runs}` +
                                        `(${currentBatter.innings.ballsFaced})`}</h5>
                                    {currentBatter.name === batter.name && <i className="fa fa-circle" />}
                                </span>
                            </div>
                        </div>
                    ))}
                <div className="row">
                    <div className="col-6">
                        <h5>{bowler.name}</h5>
                    </div>
                    <div className="col-6">
                        <h5>
                            {`${oversDescription(bowler.completedOvers, [])}-` +
                                `${bowler.maidenOvers}-${bowler.runs}-${bowler.wickets}`}
                        </h5>
                    </div>
                </div>
            </div>
            <div className="col-12 col-lg-6">
                <EntryPanel ballFunctions={ballFunctions} />
            </div>
        </div>
    </div>
);
