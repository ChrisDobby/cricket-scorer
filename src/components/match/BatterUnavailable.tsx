import * as React from 'react';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogActions from '@material-ui/core/DialogActions';
import Button from '@material-ui/core/Button';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import { UnavailableReason, Innings } from '../../domain';

interface BatterUnavailableProps {
    reason: UnavailableReason;
    innings: Innings;
    battingPlayers: string[];
    update: (playerIndex: number, reason: UnavailableReason) => void;
    cancel: () => void;
}

export default (props: BatterUnavailableProps) => {
    const [playerIndex, setPlayerIndex] = React.useState(undefined as number | undefined);

    const batters = props.innings.batting.batters.filter(
        batter =>
            (props.reason === UnavailableReason.Absent && !batter.innings) ||
            (props.reason === UnavailableReason.Retired && batter.innings && !batter.innings.wicket),
    );
    const description = props.reason === UnavailableReason.Retired ? 'retired' : 'absent';

    return (
        <div>
            <Dialog open={true} aria-labelledby="update-innings-overs-title">
                <DialogTitle id="update-innings-overs-title">{`Mark one of the batting side as ${description}`}</DialogTitle>
                <DialogContent>
                    <DialogContentText>{`Select the ${description} player`}</DialogContentText>
                </DialogContent>
                <DialogContent>
                    {batters.length > 0 && (
                        <Select value={playerIndex} onChange={ev => setPlayerIndex(Number(ev.target.value))} fullWidth>
                            {batters.map(batter => (
                                <MenuItem key={batter.playerIndex} value={batter.playerIndex}>
                                    {props.battingPlayers[batter.playerIndex]}
                                </MenuItem>
                            ))}
                        </Select>
                    )}
                    {batters.length === 0 && (
                        <DialogContentText>{`No players can be marked as ${description}`}</DialogContentText>
                    )}
                </DialogContent>
                {batters.length > 0 && (
                    <DialogActions>
                        <Button
                            onClick={() => props.update(playerIndex as number, props.reason)}
                            color="primary"
                            autoFocus
                            disabled={typeof playerIndex === 'undefined'}
                        >
                            OK
                        </Button>
                        <Button onClick={props.cancel} color="primary">
                            Cancel
                        </Button>
                    </DialogActions>
                )}
                {batters.length === 0 && (
                    <DialogActions>
                        <Button onClick={props.cancel} color="primary" autoFocus>
                            OK
                        </Button>
                    </DialogActions>
                )}
            </Dialog>
        </div>
    );
};
