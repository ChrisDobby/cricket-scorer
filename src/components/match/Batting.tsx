import * as React from 'react';
import { Extras } from './Extras';
import { Batting as InningsBatting, BattingInnings, howOutDescription } from '../../domain';
import * as styles from './styles';

const battingContainer: React.CSSProperties = {
    width: '100%',
    padding: '20px',
    marginBottom: '20px',
};

const howOut = (innings?: BattingInnings): string => {
    if (!innings) { return ''; }

    return howOutDescription(innings.wicket);
};

interface InningsItemProps { innings?: BattingInnings; }
const Howout = (props: InningsItemProps) => (
    <div className="col-md-4">{howOut(props.innings)}</div>
);

interface CellItemProps {
    property: string;
    style: React.CSSProperties;
}
const CellItem = ({ property, style }: CellItemProps) => (
    <div className="col-md-1" style={style}>{property}</div>
);

export interface BattingProps { batting: InningsBatting; }

export const Batting = ({ batting }: BattingProps) => (
    <div className="col-xl-8 col-lg-12">
        <div style={battingContainer}>
            <div className="row" style={styles.headingRow}>
                <div className="col-md-7"><h6>Batsman</h6></div>
                <div className="col-md-1" style={styles.numberCell}><h6>Runs</h6></div>
                <div className="col-md-1" style={styles.numberCell}><h6>Balls</h6></div>
                <div className="col-md-1" style={styles.numberCell}><h6>Mins</h6></div>
                <div className="col-md-1" style={styles.numberCell}><h6>4s</h6></div>
                <div className="col-md-1" style={styles.numberCell}><h6>6s</h6></div>
            </div>    
            {batting.batters.map(batter =>
                <div className="row" style={styles.itemRow} key={batter.position}>
                    <div className="col-md-3">{batter.name}</div>
                    <Howout innings={batter.innings} />
                    <CellItem property={batter.innings ? batter.innings.runs.toString() : ''} style={styles.runsCell} />
                    <CellItem
                        property={batter.innings ? batter.innings.ballsFaced.toString() : ''}
                        style={styles.numberCell}
                    />
                    <CellItem property={batter.innings ? '0' : ''} style={styles.numberCell} />
                    <CellItem
                        property={batter.innings ? batter.innings.fours.toString() : ''}
                        style={styles.numberCell}
                    />
                    <CellItem
                        property={batter.innings ? batter.innings.sixes.toString() : ''}
                        style={styles.numberCell}
                    />
                </div>)}
            <Extras extras={batting.extras} />
        </div>
    </div>
);
