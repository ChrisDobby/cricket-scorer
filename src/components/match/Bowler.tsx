import * as React from 'react';
import * as styles from './styles';
import * as globalStyles from '../styles';

const indicatorStyle: React.CSSProperties = {
    ...globalStyles.primaryColour,
    float: 'left',
    width: '30px',
    textAlign: 'center',
};

const nonSelectableStyle: React.CSSProperties = {
    color: 'rgb(233, 236, 239)',
};

interface SelectionIndicatorProps {
    playerIndex: number;
    selected: boolean;
}

const SelectionIndicator = ({ playerIndex, selected }: SelectionIndicatorProps) => {
    return (
        <div style={indicatorStyle}>
            {selected && <i className="fa fa-circle" />}
        </div>
    );
};

export interface BowlerProps {
    index: number;
    selected: boolean;
    name: string;
    allowed: boolean;
    selectBowler: (bowler: number) => void;
}

export const Bowler = ({ index, selected, name, allowed, selectBowler }: BowlerProps) => (
    <div
        className="row bowler-row"
        style={allowed ? styles.selectablePlayerStyle : nonSelectableStyle}
        onClick={() => { if (allowed) { selectBowler(index); } }}
    >
        <SelectionIndicator
            playerIndex={index}
            selected={selected}
        />
        <h6>{name}</h6>
    </div>
);
