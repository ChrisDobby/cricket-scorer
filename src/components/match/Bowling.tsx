import * as React from 'react';
import { Bowler, oversDescription } from '../../domain';
import * as styles from './styles';

export interface BowlingProps {
    team: string;
    bowlers: Bowler[];
}

export const Bowling = ({ team, bowlers }: BowlingProps) => (
    <div className="col-xl-8 col-lg-12">
        <div style={styles.sectionContainer}>
            <div className="row">
                <h6>{team} bowling</h6>
            </div>
            <div className="row" style={styles.headingRow}>
                <div className="col-6 col-md-5" />
                <div className="col-2" style={styles.centreCell}><h6>overs</h6></div>
                <div className="col-2 d-none d-md-block" style={styles.centreCell}><h6>maidens</h6></div>
                <div className="col-2" style={styles.centreCell}><h6>runs</h6></div>
                <div className="col-2 col-md-1" style={styles.centreCell}><h6>wkts</h6></div>
            </div>
            {bowlers.map(bowler => (
                <div className="row">
                    <div className="col-6 col-md-5">{bowler.name}</div>
                    <div className="col-2" style={styles.centreCell}>{oversDescription(bowler.balls)}</div>
                    <div className="col-2 d-none d-md-block" style={styles.centreCell}>{bowler.maidenOvers}</div>
                    <div className="col-2" style={styles.centreCell}>{bowler.runs}</div>
                    <div className="col-2 col-md-1" style={styles.centreCell}>{bowler.wickets}</div>
                </div>
            ))}
        </div>
    </div>
);
