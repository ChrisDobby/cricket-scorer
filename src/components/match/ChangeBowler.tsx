import * as React from 'react';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogActions from '@material-ui/core/DialogActions';
import Button from '@material-ui/core/Button';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import { Innings, Delivery } from '../../domain';

interface ChangeBowlerProps {
    innings: Innings;
    currentPlayerIndex: number;
    overNumber: number;
    currentDelivery: number;
    bowlingPlayers: string[];
    changeForFullOver: (playerIndex: number) => void;
    changeFromNow: (playerIndex: number) => void;
    changeFromDelivery: (fromDelivery: number, playerIndex: number) => void;
    cancel: () => void;
}

export default (props: ChangeBowlerProps) => {
    const invalidPlayerIndex = React.useRef(
        props.innings.events
            .filter(ev => (ev as Delivery).overNumber === props.overNumber - 1)
            .map(ev => props.innings.bowlers[(ev as Delivery).bowlerIndex].playerIndex),
    );

    const [playerIndex, setPlayerIndex] = React.useState(props.currentPlayerIndex);
    const [fullOver, setFullOver] = React.useState(props.currentDelivery > 1);
    const [fromNow, setFromNow] = React.useState(props.currentDelivery === 1);
    const [fromDelivery, setFromDelivery] = React.useState(props.currentDelivery);

    const okClick = () => {
        if (fullOver) {
            props.changeForFullOver(playerIndex);
        } else if (fromNow) {
            props.changeFromNow(playerIndex);
        } else {
            props.changeFromDelivery(fromDelivery, playerIndex);
        }
    };

    const selectFullOver = () => {
        setFullOver(true);
        setFromNow(false);
    };

    const selectFromNow = () => {
        setFullOver(false);
        setFromNow(true);
    };

    const selectFromDelivery = () => {
        setFullOver(false);
        setFromNow(false);
    };

    return (
        <div>
            <Dialog open={true} aria-labelledby="change-bowler-title">
                <DialogTitle id="change-bowler-title">{`Change the bowler of over ${props.overNumber}`}</DialogTitle>
                <DialogContent>
                    <DialogContentText>{'Select the bowler'}</DialogContentText>
                </DialogContent>
                <DialogContent>
                    <Select value={playerIndex} onChange={ev => setPlayerIndex(Number(ev.target.value))} fullWidth>
                        {props.bowlingPlayers
                            .map((player, index) => ({
                                player,
                                index,
                            }))
                            .filter(pl => !invalidPlayerIndex.current.find(idx => idx === pl.index))
                            .map(pl => (
                                <MenuItem key={pl.index} value={pl.index}>
                                    {pl.player}
                                </MenuItem>
                            ))}
                    </Select>
                </DialogContent>
                {props.currentDelivery > 1 && (
                    <DialogContent>
                        <div>
                            <Button color="primary" disabled={fullOver} onClick={selectFullOver}>
                                Full over
                            </Button>
                            <Button color="primary" disabled={fromNow} onClick={selectFromNow}>
                                From now
                            </Button>
                            <Button color="primary" disabled={!fullOver && !fromNow} onClick={selectFromDelivery}>
                                Select ball
                            </Button>
                        </div>
                        <div style={{ height: '50px' }}>
                            {!fullOver && !fromNow && (
                                <Select
                                    value={fromDelivery}
                                    onChange={ev => setFromDelivery(Number(ev.target.value))}
                                    fullWidth
                                >
                                    {Array.from(Array(props.currentDelivery).keys())
                                        .map(x => x + 1)
                                        .map(del => (
                                            <MenuItem key={del} value={del}>
                                                {del}
                                            </MenuItem>
                                        ))}
                                </Select>
                            )}
                        </div>
                    </DialogContent>
                )}
                <DialogActions>
                    <Button onClick={okClick} color="primary" autoFocus>
                        OK
                    </Button>
                    <Button onClick={props.cancel} color="primary">
                        Cancel
                    </Button>
                </DialogActions>
            </Dialog>
        </div>
    );
};
