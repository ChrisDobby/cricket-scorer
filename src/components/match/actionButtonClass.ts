import { DeliveryOutcome } from '../../domain';

const actionButtonClass = (invalidClassName: string) => (deliveryOutcome: DeliveryOutcome) =>
    deliveryOutcome === DeliveryOutcome.Noball || deliveryOutcome === DeliveryOutcome.Wide
        ? invalidClassName
        : undefined;

export default actionButtonClass('btn-danger');
