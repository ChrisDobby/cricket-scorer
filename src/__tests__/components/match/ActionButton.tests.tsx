import * as React from 'react';
import { shallow } from 'enzyme';
import * as ReactTestRenderer from 'react-test-renderer';
import { ActionButton } from '../../../components/match/ActionButton';

describe('ActionButton', () => {
    it('should call action when clicked', () => {
        const action = jest.fn();
        const actionButton = shallow(<ActionButton caption="1" action={action} />);

        actionButton.find('button').at(0).simulate('click');

        expect(action).toHaveBeenCalled();
    });

    it('should render correctly', () => {
        const actionButton = ReactTestRenderer.create(<ActionButton caption="1" action={() => { }} />);

        expect(actionButton.toJSON()).toMatchSnapshot();
    });
});
