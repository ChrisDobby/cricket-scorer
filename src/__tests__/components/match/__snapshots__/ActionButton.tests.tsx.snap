// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton should render correctly 1`] = `
<button
  className="btn btn-success"
  onClick={[Function]}
  style={
    Object {
      "margin": "2px",
      "minWidth": "36px",
    }
  }
>
  1
</button>
`;

exports[`ActionButton should render correctly when no ball set 1`] = `
<button
  className="btn btn-success"
  onClick={[Function]}
  style={
    Object {
      "margin": "2px",
      "minWidth": "36px",
      "paddingLeft": "4px",
      "paddingRight": "4px",
    }
  }
>
  <div
    style={
      Object {
        "borderColor": "#ffffff",
        "borderRadius": "50%",
        "borderStyle": "solid",
        "borderWidth": "1px",
      }
    }
  >
    1
  </div>
</button>
`;
